File: aws-lambda/src/databricks_cdk/resources/permissions/experiment_permissions.py
from enum import Enum
from typing import List, Union

from pydantic import BaseModel

from databricks_cdk.resources.permissions.models import Group, ServicePrincipal, User
from databricks_cdk.utils import CnfResponse, put_request


class ExperimentPermission(str, Enum):
    CAN_READ = "CAN_READ"
    CAN_EDIT = "CAN_EDIT"
    CAN_MANAGE = "CAN_MANAGE"


class UserPermission(User):
    permission_level: ExperimentPermission


class GroupPermission(Group):
    permission_level: ExperimentPermission


class ServicePrincipalPermission(ServicePrincipal):
    permission_level: ExperimentPermission


class ExperimentPermissionProperties(BaseModel):
    action: str = "experiment-permission"
    experiment_id: str
    access_control_list: List[Union[UserPermission, GroupPermission, ServicePrincipalPermission]]
    workspace_url: str


def get_experiment_permissions_url(workspace_url: str, experiment_id: str):
    return f"{workspace_url}/api/2.0/permissions/experiments/{experiment_id}"


def create_or_update_experiment_permissions(properties: ExperimentPermissionProperties):
    """
    Creates and updates the permissions on an existing experiment. It overwrites any existing
    permissions
    """
    body = {"access_control_list": [a.dict() for a in properties.access_control_list]}

    put_request(f"{get_experiment_permissions_url(properties.workspace_url, properties.experiment_id)}", body)

    return CnfResponse(physical_resource_id=f"{properties.experiment_id}/permissions")


def delete_experiment_permissions(properties: ExperimentPermissionProperties, physical_resource_id: str):
    """Removes all of the permission properties from an experiment"""
    put_request(
        f"{get_experiment_permissions_url(properties.workspace_url, properties.experiment_id)}",
        {"access_control_list": []},
    )

    return CnfResponse(physical_resource_id=f"{physical_resource_id}/permissions")

File: typescript/src/resources/permissions/experiment-permissions.ts
import {CustomResource} from "aws-cdk-lib";
import {Construct} from "constructs";
import {User, Group, ServicePrincipal} from "./models";


export enum ExperimentPermissionLevel {
    "CAN_READ",
    "CAN_EDIT",
    "CAN_MANAGE"
}

export interface UserExperimentPermission extends User {
    permission_level: ExperimentPermissionLevel
}

export interface GroupExperimentPermission extends Group {
    permission_level: ExperimentPermissionLevel
}

export interface ServicePrinicpalExperimentPermission extends ServicePrincipal {
    permission_level: ExperimentPermissionLevel
}


export interface ExperimentPermissionsProperties {
    workspaceUrl: string
    experimentId: string
    accessControlList: Array<UserExperimentPermission | GroupExperimentPermission | ServicePrinicpalExperimentPermission>
}

export interface ExperimentPermissionsProps extends ExperimentPermissionsProperties {
    readonly serviceToken: string
}

export class ExperimentPermissions extends CustomResource {
    constructor(scope: Construct, id: string, props: ExperimentPermissionsProps) {
        super(scope, id, {
            serviceToken: props.serviceToken,
            properties: {
                action: "experiment-permission",
                workspace_url: props.workspaceUrl,
                experiment_id: props.experimentId,
                access_control_list: props.accessControlList,
            }
        });
    }
}
